File: python_server/signature_validator.py
import base64
from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.serialization import load_pem_public_key

class SignatureValidationError(Exception):
    pass

def validate_sypago_signature(signature_b64: str, payload: bytes, nonce: str):
    """
    Valida la firma de una notificación de SyPago.
    """
    # --- Secretos ---
    # El `operation_secret` debería ser recuperado de una base de datos o un sistema de caché.
    # Para este ejemplo, usamos el valor proporcionado.
    operation_secret = "1bf38a9d-3205-4b8e-b8d1-ce95466ceef2"

    # La `public_key_pem` debería ser obtenida del endpoint de SyPago y cacheada.
    # Para este ejemplo, usamos la clave proporcionada.
    public_key_pem = b"""-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE9hXXl4g886loHmI10dLrJWFHEB8r
cyqD1hBdIM3ekQkb5YTpOShAu+7xk0fyL/0IBjLEKwSd7rsKrYGtIgJj0w==
-----END PUBLIC KEY-----"""

    # --- Proceso de Verificación ---
    
    # 1. Reconstruir el mensaje a verificar (payload.nonce.operationSecret)
    string_to_verify = b'.'.join([
        payload,
        nonce.encode('utf-8'),
        operation_secret.encode('utf-8')
    ])

    try:
        # 2. Cargar la clave pública PEM
        public_key = load_pem_public_key(public_key_pem, backend=None)

        # 3. Decodificar la firma Base64
        signature_bytes = base64.b64decode(signature_b64)
        
        # 4. Verificar la firma usando el hash SHA-256 y la curva elíptica apropiada.
        public_key.verify(
            signature_bytes,
            string_to_verify,
            ec.ECDSA(hashes.SHA256())
        )
        
        # Si la verificación es exitosa, la función no retorna nada.
        
    except InvalidSignature:
        raise SignatureValidationError("La firma no es válida")
    except Exception as e:
        # Captura otros posibles errores, como un mal formato de la clave o firma.
        raise SignatureValidationError(f"Error en el proceso de validación: {e}") 

File: js_server/signature-validator.js
const crypto = require('crypto');

/**
 * Valida la firma de una notificación de SyPago.
 * @param {string} signatureB64 Firma en formato Base64 recibida en el header X-Signature.
 * @param {Buffer} payloadBuffer El cuerpo crudo (raw) de la solicitud.
 * @param {string} nonce El nonce recibido en el header X-Signature-Nonce.
 * @returns {boolean}
 * @throws {Error} Si la firma no es válida.
 */
function validateSyPagoSignature(signatureB64, payloadBuffer, nonce) {
    // --- Secretos ---
    // El `operationSecret` debería ser recuperado de su base de datos.
    const operationSecret = '1bf38a9d-3205-4b8e-b8d1-ce95466ceef2';

    // La `publicKeyPEM` debería ser obtenida del endpoint de SyPago y cacheada.
    const publicKeyPEM = `-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE9hXXl4g886loHmI10dLrJWFHEB8r
cyqD1hBdIM3ekQkb5YTpOShAu+7xk0fyL/0IBjLEKwSd7rsKrYGtIgJj0w==
-----END PUBLIC KEY-----`;

    // --- Proceso de Verificación ---

    // 1. Reconstruir el mensaje a verificar (payload.nonce.operationSecret)
    // Se concatena el buffer del payload con los demás datos para evitar problemas de codificación.
    const stringToVerify = Buffer.concat([
        payloadBuffer,
        Buffer.from(`.${nonce}.${operationSecret}`)
    ]);

    // 2. Crear un objeto de verificación con el algoritmo SHA-256
    const verify = crypto.createVerify('SHA256');

    // 3. Cargar el mensaje en el objeto
    verify.update(stringToVerify);
    verify.end();

    // 4. Verificar la firma usando la clave pública y la firma en Base64.
    // El formato esperado por crypto.verify (ASN.1 DER) es el que se usa.
    const isSignatureValid = verify.verify(publicKeyPEM, signatureB64, 'base64');

    if (!isSignatureValid) {
        throw new Error('La firma no es válida');
    }

    return true;
}

module.exports = { validateSyPagoSignature };